The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11845389524/job/33010780480) [(plain)](https://github.com/rust-lang/rust/commit/f112e9a15521eac7dc681deb4024bf64ca3f0e8b/checks/33010780480/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
...........i....ii.i.i.i...ii.................

failures:

---- [codegen] tests/codegen/float/f16-f128-inline.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-18/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll" "/checkout/tests/codegen/float/f16-f128-inline.rs" "--check-prefix=CHECK" "--check-prefix" "NONMSVC" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
--- stderr -------------------------------
/checkout/tests/codegen/float/f16-f128-inline.rs:30:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_ref_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:17:20: note: found here
; f16_f128_inline::f16_ref_arg
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:35:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_ref_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:26:20: note: found here
; f16_f128_inline::f16_ref_ret
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:40:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_ref_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:35:20: note: found here
; f16_f128_inline::f128_ref_arg
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:45:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_ref_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:44:20: note: found here
; f16_f128_inline::f128_ref_ret
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:50:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_mut_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:53:20: note: found here
; f16_f128_inline::f16_mut_arg
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:55:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f16_mut_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:62:20: note: found here
; f16_f128_inline::f16_mut_ret
                   ^~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:60:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_mut_arg
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:71:20: note: found here
; f16_f128_inline::f128_mut_arg
                   ^~~~~~~~~~~~
/checkout/tests/codegen/float/f16-f128-inline.rs:65:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: f128_mut_ret
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll:80:20: note: found here
; f16_f128_inline::f128_mut_ret


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/float/f16-f128-inline/f16-f128-inline.ll
Check file: /checkout/tests/codegen/float/f16-f128-inline.rs

-dump-input=help explains the following input dump.
Input was:
<<<<<<
<<<<<<
        1: ; ModuleID = 'f16_f128_inline.59ce32246df2b6e0-cgu.0' 
        2: source_filename = "f16_f128_inline.59ce32246df2b6e0-cgu.0" 
        3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
        4: target triple = "x86_64-unknown-linux-gnu" 
        5:  
        6: @alloc_aa07815cbcb2365f7aca41cc8941a0c4 = private unnamed_addr constant <{ [19 x i8] }> <{ [19 x i8] c"not yet implemented" }>, align 1 
        7: @alloc_6bc582bffdc44de442aab6f66dd365b4 = private unnamed_addr constant <{ [48 x i8] }> <{ [48 x i8] c"/checkout/tests/codegen/float/f16-f128-inline.rs" }>, align 1 
        8: @alloc_82b2a2cbad8c28ba4564dc4ecbd7e9bc = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00\1F\00\00\00" }>, align 8 
        9: @alloc_60f1c4e6531fb6b6d90525e37dd9aa14 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00$\00\00\00" }>, align 8 
       10: @alloc_79cc007c1b095e2bdc6b96ef7d0bdf0c = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00)\00\00\00" }>, align 8 
       11: @alloc_818ee584f1ec5ee88ba376a8cbcb7c4e = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00.\00\00\00" }>, align 8 
       12: @alloc_4327bb663ca54a7b8c119379e0f05c48 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\003\00\00\00" }>, align 8 
       13: @alloc_0d04e59b3c37a8041aee1430f660b11f = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\008\00\00\00" }>, align 8 
       14: @alloc_a890b3d25fe4f6b1d177a274990e6b06 = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00=\00\00\00" }>, align 8 
       15: @alloc_3c1d6e3ca54baaac102563202b3e6dbb = private unnamed_addr constant <{ ptr, [16 x i8] }> <{ ptr @alloc_6bc582bffdc44de442aab6f66dd365b4, [16 x i8] c"0\00\00\00\00\00\00\00\05\00\00\00B\00\00\00" }>, align 8 
       17: ; f16_f128_inline::f16_ref_arg 
       17: ; f16_f128_inline::f16_ref_arg 
not:30                        !~~~~~~~~~~  error: no match expected
       18: ; Function Attrs: noreturn nonlazybind uwtable 
       19: define void @_ZN15f16_f128_inline11f16_ref_arg17hc34aff3c97540dd2E(ptr noalias nocapture noundef readonly align 2 dereferenceable(2) %_a) unnamed_addr #0 { 
       21: ; call core::panicking::panic 
       21: ; call core::panicking::panic 
       22:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_82b2a2cbad8c28ba4564dc4ecbd7e9bc) #2 
       23:  unreachable 
       24: } 
       25:  
       26: ; f16_f128_inline::f16_ref_ret 
not:35                        !~~~~~~~~~~  error: no match expected
       27: ; Function Attrs: noreturn nonlazybind uwtable 
       28: define noalias noundef nonnull align 2 dereferenceable(2) ptr @_ZN15f16_f128_inline11f16_ref_ret17h70d29570d32add63E() unnamed_addr #0 { 
       30: ; call core::panicking::panic 
       30: ; call core::panicking::panic 
       31:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_60f1c4e6531fb6b6d90525e37dd9aa14) #2 
       32:  unreachable 
       33: } 
       35: ; f16_f128_inline::f128_ref_arg 
       35: ; f16_f128_inline::f128_ref_arg 
not:40                        !~~~~~~~~~~~  error: no match expected
       36: ; Function Attrs: noreturn nonlazybind uwtable 
       37: define void @_ZN15f16_f128_inline12f128_ref_arg17h899b1200f7e59c3cE(ptr noalias nocapture noundef readonly align 16 dereferenceable(16) %_a) unnamed_addr #0 { 
       39: ; call core::panicking::panic 
       39: ; call core::panicking::panic 
       40:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_79cc007c1b095e2bdc6b96ef7d0bdf0c) #2 
       41:  unreachable 
       42: } 
       43:  
       44: ; f16_f128_inline::f128_ref_ret 
not:45                        !~~~~~~~~~~~  error: no match expected
       45: ; Function Attrs: noreturn nonlazybind uwtable 
       46: define noalias noundef nonnull align 16 dereferenceable(16) ptr @_ZN15f16_f128_inline12f128_ref_ret17h75f4b575024e18bcE() unnamed_addr #0 { 
       48: ; call core::panicking::panic 
       48: ; call core::panicking::panic 
       49:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_818ee584f1ec5ee88ba376a8cbcb7c4e) #2 
       50:  unreachable 
       51: } 
       52:  
       53: ; f16_f128_inline::f16_mut_arg 
not:50                        !~~~~~~~~~~  error: no match expected
       54: ; Function Attrs: noreturn nonlazybind uwtable 
       55: define void @_ZN15f16_f128_inline11f16_mut_arg17he179779c24a59fc0E(ptr noalias nocapture noundef readnone align 2 dereferenceable(2) %_a) unnamed_addr #0 { 
       57: ; call core::panicking::panic 
       57: ; call core::panicking::panic 
       58:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_4327bb663ca54a7b8c119379e0f05c48) #2 
       59:  unreachable 
       60: } 
       61:  
       62: ; f16_f128_inline::f16_mut_ret 
not:55                        !~~~~~~~~~~  error: no match expected
       63: ; Function Attrs: noreturn nonlazybind uwtable 
       64: define noalias noundef nonnull align 2 dereferenceable(2) ptr @_ZN15f16_f128_inline11f16_mut_ret17h7e5123d644701330E() unnamed_addr #0 { 
       66: ; call core::panicking::panic 
       66: ; call core::panicking::panic 
       67:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_0d04e59b3c37a8041aee1430f660b11f) #2 
       68:  unreachable 
       69: } 
       70:  
       71: ; f16_f128_inline::f128_mut_arg 
not:60                        !~~~~~~~~~~~  error: no match expected
       72: ; Function Attrs: noreturn nonlazybind uwtable 
       73: define void @_ZN15f16_f128_inline12f128_mut_arg17h2020b9428d06f7eaE(ptr noalias nocapture noundef readnone align 16 dereferenceable(16) %_a) unnamed_addr #0 { 
       75: ; call core::panicking::panic 
       75: ; call core::panicking::panic 
       76:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_a890b3d25fe4f6b1d177a274990e6b06) #2 
       77:  unreachable 
       78: } 
       79:  
       80: ; f16_f128_inline::f128_mut_ret 
not:65                        !~~~~~~~~~~~  error: no match expected
       81: ; Function Attrs: noreturn nonlazybind uwtable 
       82: define noalias noundef nonnull align 16 dereferenceable(16) ptr @_ZN15f16_f128_inline12f128_mut_ret17h512746cbe2a0f388E() unnamed_addr #0 { 
       84: ; call core::panicking::panic 
       84: ; call core::panicking::panic 
       85:  tail call void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1 @alloc_aa07815cbcb2365f7aca41cc8941a0c4, i64 noundef 19, ptr noalias noundef nonnull readonly align 8 dereferenceable(24) @alloc_3c1d6e3ca54baaac102563202b3e6dbb) #2 
       86:  unreachable 
       87: } 
       89: ; core::panicking::panic 
       89: ; core::panicking::panic 
       90: ; Function Attrs: cold noinline noreturn nonlazybind uwtable 
       91: declare void @_RNvNtCs5y47RM21p7k_4core9panicking5panic(ptr noalias noundef nonnull readonly align 1, i64 noundef, ptr noalias noundef readonly align 8 dereferenceable(24)) unnamed_addr #1 
       92:  
       93: attributes #0 = { noreturn nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       94: attributes #1 = { cold noinline noreturn nonlazybind uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
       95: attributes #2 = { noreturn } 
       96:  
       97: !llvm.module.flags = !{!0, !1} 
       98: !llvm.ident = !{!2} 
       99:  
      100: !0 = !{i32 8, !"PIC Level", i32 2} 
      101: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
      102: !2 = !{!"rustc version 1.84.0-nightly (16bf38087 2024-11-14)"} 
------------------------------------------



```

</details>

--- a/compiler/rustc_mir_transform/src/cross_crate_inline.rs
+++ b/compiler/rustc_mir_transform/src/cross_crate_inline.rs
@@ -50,6 +50,16 @@ fn cross_crate_inlinable(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
         _ => {}
     }
 
+    let sig = tcx.fn_sig(def_id).instantiate_identity();
+    for ty in sig.inputs().skip_binder().iter().chain(std::iter::once(&sig.output().skip_binder()))
+    {
+        // FIXME(f16_f128): in order to avoid crashes building `core`, always inline to skip
+        // codegen if the function is not used.
+        if ty == &tcx.types.f16 || ty == &tcx.types.f128 {
+            return true;
+        }
+    }
+
     // Don't do any inference when incremental compilation is enabled; the additional inlining that
     // inference permits also creates more work for small edits.
     if tcx.sess.opts.incremental.is_some() {

--- a/library/core/src/num/f128.rs
+++ b/library/core/src/num/f128.rs
@@ -1258,9 +1258,8 @@ impl f128 {
             min <= max,
             "min > max, or either was NaN",
             "min > max, or either was NaN. min = {min:?}, max = {max:?}",
-            // FIXME(f16_f128): Passed by-ref to avoid codegen crashes
-            min: &f128 = &min,
-            max: &f128 = &max,
+            min: f128,
+            max: f128,
         );
 
         if self < min {

--- a/library/core/src/num/f16.rs
+++ b/library/core/src/num/f16.rs
@@ -1235,9 +1235,8 @@ impl f16 {
             min <= max,
             "min > max, or either was NaN",
             "min > max, or either was NaN. min = {min:?}, max = {max:?}",
-            // FIXME(f16_f128): Passed by-ref to avoid codegen crashes
-            min: &f16 = &min,
-            max: &f16 = &max,
+            min: f16,
+            max: f16,
         );
 
         if self < min {

--- a/tests/codegen/float/f16-f128-inline.rs
+++ b/tests/codegen/float/f16-f128-inline.rs
@@ -1,0 +1,29 @@
+//@ revisions: default nopt
+//@[nopt] compile-flags: -Copt-level=0 -Zcross-crate-inline-threshold=never -Zmir-opt-level=0 -Cno-prepopulate-passes
+
+// Ensure that functions using `f16` and `f128` are always inlined to avoid crashes
+// when the backend does not support these types.
+
+#![crate_type = "lib"]
+#![feature(f128)]
+#![feature(f16)]
+
+pub fn f16_arg(_a: f16) {
+    // CHECK-NOT: f16_arg
+    todo!()
+}
+
+pub fn f16_ret() -> f16 {
+    // CHECK-NOT: f16_ret
+    todo!()
+}
+
+pub fn f128_arg(_a: f128) {
+    // CHECK-NOT: f128_arg
+    todo!()
+}
+
+pub fn f128_ret() -> f128 {
+    // CHECK-NOT: f128_ret
+    todo!()
+}

